stablehlo/tests/stablehlo_aggressive_simplification.mlir

```diff
@@ -664,6 +664,237 @@ func.func @reduce_zero_ext(%arg0: tensor<0xi1>) -> tensor<i32> {
+  // CHECK: [[R0:%.+]] = stablehlo.constant dense<1.{{0+}}e+00> : tensor<f32>
+  // CHECK: [[R1:%.+]] = stablehlo.reduce(%arg0 init: [[R0]]) applies stablehlo.add
+  %0 = stablehlo.constant dense<1.0> : tensor<f32>
+  %1 = stablehlo.constant dense<2> : tensor<i64>
+  %2 = stablehlo.constant dense<3> : tensor<i64>
+  %3:3 = stablehlo.reduce(%arg0 init: %0), (%arg1 init: %1), (%arg2 init: %2) across dimensions = [0] :
+  (tensor<8xf32>, tensor<8xi64>, tensor<8xi64>, tensor<f32>, tensor<i64>, tensor<i64>) ->
+  (tensor<f32>, tensor<i64>, tensor<i64>)
+   reducer(%arg3: tensor<f32>, %arg6: tensor<f32>)
+          (%arg4: tensor<i64>, %arg7: tensor<i64>)
+          (%arg5: tensor<i64>, %arg8: tensor<i64>)
+  {
+    %4 = stablehlo.add %arg3, %arg6 : tensor<f32>
+    %5 = stablehlo.minimum %arg3, %arg6 : tensor<f32>
+    %6 = stablehlo.convert %5 : (tensor<f32>) -> (tensor<i64>)
+    %8 = stablehlo.maximum %arg5, %arg8 : tensor<i64>
+    stablehlo.return %4, %8, %6 : tensor<f32>, tensor<i64>, tensor<i64>
+  }
+  return %3#0 : tensor<f32>
+}
+
+// -----
+
+//  (a, a1) (b, b1) (c, c1)
+//    ▴       ▴       ▴
+//    │       │       │
+//    │       ├──────►│
+//    │       │◄──────┤
+//    │       │       │
+//    r0      r1      r2
+//    U
+//
+// Both r1 and r2 can be dropped despite circular dependency.
+
+// CHECK-LABEL: func.func @reduce_unused_case5
+func.func @reduce_unused_case5(%arg0: tensor<8xf32>,
+                               %arg1: tensor<8xi64>,
+                               %arg2: tensor<8xi64>) -> tensor<f32> {
+  // CHECK: [[R0:%.+]] = stablehlo.constant dense<1.{{0+}}e+00> : tensor<f32>
+  // CHECK: [[R1:%.+]] = stablehlo.reduce(%arg0 init: [[R0]]) applies stablehlo.add
+  %0 = stablehlo.constant dense<1.0> : tensor<f32>
+  %1 = stablehlo.constant dense<2> : tensor<i64>
+  %2 = stablehlo.constant dense<3> : tensor<i64>
+  %3:3 = stablehlo.reduce(%arg0 init: %0), (%arg1 init: %1), (%arg2 init: %2) across dimensions = [0] :
+  (tensor<8xf32>, tensor<8xi64>, tensor<8xi64>, tensor<f32>, tensor<i64>, tensor<i64>) ->
+  (tensor<f32>, tensor<i64>, tensor<i64>)
+   reducer(%arg3: tensor<f32>, %arg6: tensor<f32>)
+          (%arg4: tensor<i64>, %arg7: tensor<i64>)
+          (%arg5: tensor<i64>, %arg8: tensor<i64>)
+  {
+    %4 = stablehlo.add %arg3, %arg6 : tensor<f32>
+    %5 = stablehlo.add %arg4, %arg7 : tensor<i64>
+    %6 = stablehlo.add %arg5, %arg8 : tensor<i64>
+    %7 = stablehlo.minimum %5, %6 : tensor<i64>
+    %8 = stablehlo.maximum %5, %6 : tensor<i64>
+    stablehlo.return %4, %7, %8 : tensor<f32>, tensor<i64>, tensor<i64>
+  }
+  return %3#0 : tensor<f32>
+}
+
+// -----
+
+//  (a, a1) (b, b1) (c, c1)
+//    ▴       ▴       ▴
+//    │       ├──────►│
+//    │       │◄──────│
+//    │◄──────┼───────┤
+//    │       │       │
+//    r0      r1      r2
+//            U
+//
+// Non-conservative case. In theory r0 can be removed.
+// But what should be done with r2?
```

Comment:
Hm, i think arrows are a bit misleading. I was thinking about an arrow as a abstract set of instructions that is, while traversed through def-use chain, brings us to block arguments that are used to compute return operand.
These sets can definitely intersect, as one instruction can be used to calculate multiple return results/operands.
```
...
   reducer(%arg3: tensor<f32>, %arg6: tensor<f32>)
          (%arg4: tensor<i64>, %arg7: tensor<i64>)
          (%arg5: tensor<i64>, %arg8: tensor<i64>)
  {
    %4 = stablehlo.add %arg3, %arg6 : tensor<f32>
    %5 = stablehlo.add %arg4, %arg7 : tensor<i64>
    %6 = stablehlo.add %arg5, %arg8 : tensor<i64>
    %7 = stablehlo.minimum %5, %6 : tensor<i64>
    %8 = stablehlo.maximum %5, %6 : tensor<i64>
    %9 = stablehlo.convert %4 : (tensor<f32>) -> (tensor<i64>)
    %10 = stablehlo.subtract %8, %9 : tensor<i64>
    stablehlo.return %4, %7, %10 : tensor<f32>, tensor<i64>, tensor<i64>
  }
...
```
format: `result alias == result operand -> { used instruction } -> { used block arguments }`
r0 == %4 -> { %4 } -> (%arg3, %arg6)
r1 == %7 -> { %7, %5, %6 } -> (%arg4, %arg7), (%arg5, %arg8)
r2 == %10 -> { %10, %8, %9, %5, %6, %4 } -> (%arg3, %arg6), (%arg4, %arg7), (%arg5, %arg8)

Its might be better to say that arrow is `{ set of used instructions }` plus `{ set of used block arguments }`.

In example above `r0` and `r2` are unused, so they their arrows can be dropped, except those elements that are present in `r1` arrow, i.e. while removing `r2` we can't delete instruction `%5, %6` and arguments `(%arg4, %arg7), (%arg5, %arg8)`. Initially i started with an approach that removes unused instructions/arguments, but it failed on cases with circular dependencies like `reduce_unused_case5` (requires optimistic erasing). Current approach is much easier, just mark what instructions and arguments should stay. In the end only one constraint `size(inputs) = size(init_values) = size(results)` must be satisfied.

So, once everything related to to `r0` and `r2` is dropped, we are left in situation
```
//  (b, b1) (c, c1)
//    ▴       ▴
//    ├───────┘
//    │
//    │
//    │
//    r1      ???
//    U
```
In theory we can place `0` instead of `???`, but i can imaging some weird lowering where operand's memref still modified in place, therefore that buffer would zero out.

Reply:
I'm not sure I follow completely. This makes sense to me: `I was thinking about an arrow as a abstract set of instructions that is, while traversed through def-use chain, brings us to block arguments that are used to compute return operand.`, that is also how I understood the arrows.

You mention: `In example above r0 and r2 are unused, so they their arrows can be dropped, except those elements that are present in r1 arrow`; IMO this is not completely accurate. Let me expand on what I said earlier about multiple iterations

In the below example (`reduce_unused_case6`), in the first iteration we have the following dependencies: `r0 --> {a, a1}`, `r1 --> {b, b1, c, c1}` and `r2 --> {a, a1, c, c1}`. However, in the second iteration, `a1`'s value is actually `r0`, `b1`'s value is `r1`, and `c1`'s value is `r2`, so we have `r1 --> {r2}` and `r2 --> {r0, r1}`. So `r0` ends up being used transitively, across iterations.

```
(a, a1) (b, b1) (c, c1)
  ▴       ▴       ▴
  │       ├──────►│
  │       │◄──────│
  │◄──────┼───────┤
  │       │       │
  r0      r1      r2
          U
```

(Note that I think the approach proposed in the PR is fine, just that we may be able to delete even more -- as I said though, that can be future work.)

Reply:
My bad, i thought something wasn't clear in my description, but it seems that i didn't grasp your point. Can you elaborate what you mean by `fist iteration`  and `second iteration`?
```
 reducer(%arg3: tensor<i64>, %arg6: tensor<i64>)
        (%arg4: tensor<i64>, %arg7: tensor<i64>)
        (%arg5: tensor<i64>, %arg8: tensor<i64>)
{
  %4 = stablehlo.add %arg3, %arg6 : tensor<i64>
  %5 = stablehlo.subtract %arg4, %arg7 : tensor<i64>
  %6 = stablehlo.multiply %arg5, %arg8 : tensor<i64>
  %7 = stablehlo.minimum %5, %6 : tensor<i64>
  %8 = stablehlo.maximum %5, %6 : tensor<i64>
  %9 = stablehlo.subtract %8, %4 : tensor<i64>
  stablehlo.return %4, %7, %9 : tensor<i64>, tensor<i64>, tensor<i64>
}
```
```
  r0 == %4 -> { %4 } -> { (%arg3, %arg6) }
U r1 == %7 -> { %7, %5, %6 } -> { (%arg4, %arg7), (%arg5, %arg8) }
  r2 == %9 -> { %9, %8, %5, %6, %4 } -> { (%arg3, %arg6), (%arg4, %arg7), (%arg5, %arg8) }
```
I see, a mistake, we can't safely remove `r0` along with `(%arg3, %arg6)` because r2 is still using `(%arg3, %arg6)`, if that is what you have meant? An idea of removing arrow of `r2` (along with releasing all its dependencies) and setting `0` as the return value led me to mistakenly believe that `r0` could be removed.

BTW, I am really interested in your opinion about setting a synthetic value like `0` as the return value to satisfy `size(inputs) = size(init_values) = size(results)` constraint. This would allow to support all possible cases.

Reply:
Thanks, your updated example is how I understood it.

By iterations, I mean e.g. that computing `reduce(+, [1, 2, 3] 0)` is the same as `(((0 + 1) + 2) + 3)` (or some other order), which takes several steps (or iterations). Here the reduction has a single argument and a single initial value, but the idea is the same. Given this example, in the first "iteration", the "initial value" is 0, and the current value is 1. This produces a result of 1. For the second iteration, the "initial value" is 1 (the previous result), and the current value is 2. And so on.

So if a result depends on an argument/initial value, it also depends on any values that the corresponding result depends on (e.g. if `r2` depends on `a0` and `r0` depends on `a1`, then `r2` also depends on `a1` transitively through `r0`).

I actually made a typo earlier when I wrote `In the below example (reduce_unused_case6), in the first iteration we have the following dependencies: r0 --> {a, a1}, r1 --> {b, b1, c, c1} and r2 --> {a, a1, c, c1}`. In fact, `r2` also depends on `{b, b1}`, I missed the arrow.

I think expanding the diagram across 2 iterations helps:

```
(a, a1) (b, b1) (c, c1)
  ▴       ▴       ▴
  │       ├──────►│
  │       │◄──────│
  │◄──────┼───────┤
  │       │       │
(a, a1) (b, b1) (c, c1)
  ▴       ▴       ▴
  │       ├──────►│
  │       │◄──────│
  │◄──────┼───────┤
  │       │       │
  r0      r1      r2
          U
```

If you follow the arrows from `r1`, you eventually reach `{a, b, c}`, so everything must be kept.

Reply:
Thanks. That's interesting, i haven't even considered this operator's semantic point of view. Does it also mean that return results a implicitly `hardwired` to input operands by design? For example:
```
(a, a1) (b, b1)
  ▴            
  │            
  │◄──────┐    
  │       │    
  │       │    
  r0      r1
  U       U    
```
 if `r1` only depends on `(a, a1)` that is `(1, 0)` on first iteration, does it also mean that on second iteration `b1` would be equal to `r1` from first the iteration, i.e. `0 + 1` even though it haven't participated in computations at all?

Reply:
As far as my understanding goes, that is correct.

Given a reducer like (using pseudo-MLIR syntax, with inlining):

```
func (%ai, %a1, %bi, %b1) {
  return %ai + %a1, %ai + %a1
}
```

On the first iteration,`%ai` is equal to the first value in the input tensor `%a` and `%a1` is equal to the initial value used to reduce `%a`.

The first iteration does: `%r0 = %a[0] + %a1, %r1 = %a[0] + %a1`.

On the second iteration, `%ai` is equal to the second value in the input tensor `%a` and `%a1` is equal to `%r0` from the previous iteration. Similarly for `%bi` and `%b1` (which is equal to `%r1`, which in this specific case is equal to `%r0`).

The second iteration does: %r0 = %a[1] + %r0, %1 = %a[1] + %r0`, and so on.

(Note that in the above I am assuming a specific right-to-left iteration order, but the `reduce` op does not make any guarantees w.r.t execution order.)

This case is interesting because `(b, b1)` is never used, but because `r1` is used, we can't just remove the argument, because the number of input pairs much match the number of results. We would need a deeper analysis to show that `r0` and `r1` are in fact computing the same thing (trivial in this case, but likely not trivial in general) before we can conclude that we can remove the `(b, b1)` argument pair and merge `r0` and `r1`.